src/librustc_yk_sections/mir_cfg.rs

```diff
@@ -64,43 +150,192 @@ pub fn generate_tir<'a, 'tcx, 'gcx>(
     Ok(ret)
 }
 
+/// Lazy computation of dominance frontiers.
+///
+/// See Page 404 of the 2nd edition of the Appel book mention above.
+///
+/// Since the frontier of one node may depend on the frontiers of other nodes (depending upon their
+/// relationships), we cache the frontiers so as to avoid computing things more than once.
+struct DominatorFrontiers<'a, 'tcx> {
+    /// The MIR we are working on.
+    mir: &'a Mir<'tcx>,
+    /// The dominators of the above MIR.
+    doms: Dominators<BasicBlock>,
+    /// The dominance frontiers. Lazily computed. None means as-yet uncomputed.
+    df: IndexVec<BasicBlock, Option<Vec<BasicBlock>>>,
+}
+
+impl<'a, 'tcx> DominatorFrontiers<'a, 'tcx> {
+    fn new(mir: &'a Mir<'tcx>) -> Self {
+        let num_blks = mir.basic_blocks().len();
+        let mut df = IndexVec::with_capacity(num_blks);
+        for _ in 0..num_blks {
+            df.push(None);
+        }
+
+        Self {
+            mir,
+            doms: mir.dominators(),
+            df,
+        }
+    }
+
+    /// Get the dominance frontier of the given basic block, computing it if we have not already
+    /// done so.
+    fn get(&mut self, n: BasicBlock) -> &Vec<BasicBlock> {
+        if self.df[n].is_none() {
+            // We haven't yet computed dominator frontiers for this node. Compute them.
+            let mut s: Vec<BasicBlock> = Vec::new();
+
+            // Append what Appel calls 'DF_{local}[n]'.
+            for y in self.mir.basic_blocks()[n].terminator().successors() {
+                if self.doms.immediate_dominator(*y) != n {
+                    s.push(*y);
+                }
+            }
+
+            // The second stage of the algorithm needs the children nodes in the dominator tree.
+            let mut children = Vec::new();
+            for (b, _) in self.mir.basic_blocks().iter_enumerated() {
+                let b = BasicBlock::from_u32(b.as_u32());
+                if self.doms.is_dominated_by(b, n) && b != n {
+                    children.push(b);
+                    // Force the frontier of `b` into `self.df`. Doing this here avoids a
```

Comment:
Not sure I understand the comment or why `get` does anything.

Reply:
Oh, this part of the code was horrendous to get into rust.

TLDR: `get()` has side effects.

Long answer:  if were were literally translating the algorithm from the book,  we'd have done the get in the next chunk of the algorithm, but it's impossible to get past the borrow checker. Get mutates an internal cache inside the dominator frontier struct. By mutating now, we can ensure the entry exists and in the next chunk do a regular index operation without any mutation.

The cache exists because the dominator frontier algorithm is like a big pot of simultaneous equations and without a cache we'd compute the same things over and over. It's a bit like memoising fib.

Hard to explain. Does it make sense?

Reply:
If `get` has side effects, it seems like a weird name for that function!

Reply:
We can call it whatever we like. In the book it's a regular hash-table operation.

How about `frontier()` because that's what it returns.

Reply:
That seems like a better name.